File: frontend/src/api/special-rule.ts
import api from './index'

export interface SpecialRule {
  id: number
  staff_id: number
  rule_type: string
  params: Record<string, any> | null
  effective_from: string | null
  effective_to: string | null
  reason: string | null
  created_at: string | null
  updated_at: string | null
}

export interface SpecialRuleCreate {
  staff_id: number
  rule_type: string
  params?: Record<string, any>
  effective_from?: string | null
  effective_to?: string | null
  reason?: string | null
}

export interface SpecialRuleUpdate {
  rule_type?: string
  params?: Record<string, any>
  effective_from?: string | null
  effective_to?: string | null
  reason?: string | null
}

export function getSpecialRules(params?: { staff_id?: number }): Promise<SpecialRule[]> {
  return api.get('/special-rules', { params })
}

export function getSpecialRule(id: number): Promise<SpecialRule> {
  return api.get(`/special-rules/${id}`)
}

export function createSpecialRule(data: SpecialRuleCreate): Promise<SpecialRule> {
  return api.post('/special-rules', data)
}

export function updateSpecialRule(id: number, data: SpecialRuleUpdate): Promise<SpecialRule> {
  return api.put(`/special-rules/${id}`, data)
}

export function deleteSpecialRule(id: number): Promise<any> {
  return api.delete(`/special-rules/${id}`)
}

File: backend/app/schemas/special_rule.py
from pydantic import BaseModel, field_validator
from typing import Optional
from datetime import date, datetime


class SpecialRuleCreate(BaseModel):
    """创建特殊排班规则"""
    staff_id: int
    rule_type: str
    params: dict = {}
    effective_from: Optional[date] = None
    effective_to: Optional[date] = None
    reason: Optional[str] = None

    @field_validator("staff_id")
    @classmethod
    def validate_staff_id(cls, v: int) -> int:
        if v <= 0:
            raise ValueError("人员ID必须大于0")
        return v

    @field_validator("rule_type")
    @classmethod
    def validate_rule_type(cls, v: str) -> str:
        valid_types = [
            "exclude_shift", "include_shift", "exclude_post",
            "must_pair", "exclude_date", "exclude_weekday",
        ]
        if v not in valid_types:
            raise ValueError(f"规则类型必须是以下之一: {', '.join(valid_types)}")
        return v

    @field_validator("effective_to")
    @classmethod
    def validate_effective_to(cls, v: Optional[date], info) -> Optional[date]:
        if v and info.data.get("effective_from") and v < info.data["effective_from"]:
            raise ValueError("结束日期不能早于开始日期")
        return v


class SpecialRuleUpdate(BaseModel):
    """更新特殊排班规则"""
    rule_type: Optional[str] = None
    params: Optional[dict] = None
    effective_from: Optional[date] = None
    effective_to: Optional[date] = None
    reason: Optional[str] = None

    @field_validator("rule_type")
    @classmethod
    def validate_rule_type(cls, v: Optional[str]) -> Optional[str]:
        if v is not None:
            valid_types = [
                "exclude_shift", "include_shift", "exclude_post",
                "must_pair", "exclude_date", "exclude_weekday",
            ]
            if v not in valid_types:
                raise ValueError(f"规则类型必须是以下之一: {', '.join(valid_types)}")
        return v


class SpecialRuleResponse(BaseModel):
    """特殊排班规则响应"""
    id: int
    staff_id: int
    rule_type: str
    params: Optional[dict] = None
    effective_from: Optional[date] = None
    effective_to: Optional[date] = None
    reason: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    model_config = {"from_attributes": True}
